torch/testing/_asserts.py

```diff
@@ -331,29 +368,230 @@ def assert_tensors_equal(
         raise exc
     actual, expected = _equalize_attributes(actual, expected)
 
-    exc = _check_values_equal(actual, expected)
+    if (rtol == 0.0) and (atol == 0.0):
+        exc = _check_values_equal(actual, expected)
+    else:
+        exc = _check_values_close(actual, expected, rtol=rtol, atol=atol)
+    if exc:
+        return exc
+
+    return None
+
+
+def _check_by_type(
+    actual: Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]],
+    expected: Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]],
+    check_tensors: Callable[[Tensor, Tensor], Optional[Exception]],
+) -> Optional[Exception]:
+    """Delegates tensor checking based on the inputs types.
+
+    Currently supports pairs of
+
+    - :class:`Tensor`'s,
+    - :class:`~collections.abc.Sequence`'s of :class:`Tensor`'s, and
+    - :class:`~collections.abc.Mapping`'s of :class:`Tensor`'s.
+
+    Args:
+        actual (Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]]): Actual input.
+        expected (Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]]): Expected input.
+        check_tensors (Callable[[Tensor, Tensor], Optional[Exception]]): Callable used to check if two tensors match.
+            In case they mismatch should return an :class:`Exception` with an expressive error message.
+
+    Returns:
+        (Optional[Exception]): :class:`UsageError` if the inputs types are unsupported. Additionally, any exception
+        returned by :attr:`check_tensors`.
+    """
+    # _check_are_tensors() returns nothing in case both inputs are tensors and an exception otherwise. Thus, the logic
+    # is inverted here.
+    are_tensors = not _check_are_tensors(actual, expected)
+    if are_tensors:
+        return check_tensors(cast(Tensor, actual), cast(Tensor, expected))
+
+    if isinstance(actual, collections.abc.Sequence) and isinstance(expected, collections.abc.Sequence):
+        return _check_sequence(actual, expected, check_tensors)
+    elif isinstance(actual, collections.abc.Mapping) and isinstance(expected, collections.abc.Mapping):
+        return _check_mapping(actual, expected, check_tensors)
+
+    return UsageError(
+        f"Both inputs have to be tensors, or sequences or mappings of tensors, "
+        f"but got {type(actual)} and {type(expected)} instead."
+    )
+
+
+E = TypeVar("E", bound=Exception)
+
+
+def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
```

Comment:
I wonder if the need for this function already warrants a change from all internal checks returning `exc: Optional[Exception]` to `exc_type: Optional[Type[Exception]], msg: Optional[str]`. It would make the default case a little more verbose

```python
exc_type, msg = _check_whatever()
if exc_type:
    raise exc_type(msg)
```

vs.

```python
exc = _check_whatever()
if exc:
    raise exc
```
but remove this awkward re-instantiation of the exception.

Reply:
Either seems reasonable for now; I like being verbose

Reply:
Lets keep it they way it is right now and think about it after more pressing issues in the stack are resolved.